src/match/queue.ts

```diff
@@ -0,0 +1,48 @@
+import { randomUUID } from 'crypto';
+
+export const subscribe = () => {};
+type Item = {
+  id: string;
+  joinAt: number;
+};
+
+export class Queue {
+  queue: Item[] = [];
+  subscribers: any[] = [];
+  subscribe({
+    onJoin,
+    onLeave,
+  }: {
+    onJoin: (playerID: string) => void;
+    onLeave: (playerID: string) => void;
+  }) {
+    const id = randomUUID();
+    this.subscribers.push({ id, onJoin, onLeave });
+  }
+  join(playerID: string) {
+    // TODO - check if playerID is already exist
+    this.queue.push({
+      id: playerID,
+      joinAt: Date.now(),
+    });
+    this.subscribers.forEach((subscriber) => subscriber.onJoin(playerID));
+  }
+  leave(playerID: string) {
+    this.queue = this.queue.filter(({ id }) => {
+      return id !== playerID;
+    });
+    this.subscribers.forEach((subscriber) => subscriber.onLeave(playerID));
+  }
+  get() {
+    return this.queue;
+  }
+  getFirstNItems(n: number) {
+    if (n > this.queue.length) {
+      return;
+    }
+    return this.queue.slice(0, n);
+  }
+  deleteMatchedPlayers(playerIDs: string[]) {
+    this.queue = this.queue.filter(({ id }) => !playerIDs.includes(id));
+  }
```

Comment:
So for these last three methods `get`, `getFirstNItems` and `deleteMatchedPlayers`, I think it'd be better to not expose these publicly. 

I also think the following would be a better API:

- When the caller code adds a player to the queue, it calls `join()`. 
- `join()` then returns either a null object, or a list containing the usernames that are matched.  
- When `join()` does this, it'll automatically remove those usernames from the list. This way the caller doesn't have to separately call `deleteMatchedPlayers()`. 
- When a player leaves the queue, simply call `leave()`. 

This will simplify our API and make it easier to use :). Later on when we need to matchmake with elo, we'll pass in the user's elo into the `join()` method and add that extra complexity in `join()`.

Reply:
` join(playerId: string) {`
    `this.queue.push({`
      `id: playerId,`
     ` joinAt: Date.now(),`
    `});`
    `this.subscribers.forEach((subscriber) => subscriber.onJoin(playerId));`
    `return this.queue`
  }`

Is this what you mean? every time they join, it will return the queue as well

Reply:
Nah, I was thinking it'll only return 6 usernames if there are 6 people in the queue now. If not, then return null.